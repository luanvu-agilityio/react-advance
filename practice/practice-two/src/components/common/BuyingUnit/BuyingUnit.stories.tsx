import type { Meta, StoryObj } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import { within, userEvent, expect } from '@storybook/test'
import BuyingUnit from './BuyingUnit'

const meta: Meta<typeof BuyingUnit> = {
  title: 'Components/ProductDetails/BuyingUnit',
  component: BuyingUnit,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component:
          'A component for selecting quantity and unit type for product purchases.',
      },
    },
  },
  argTypes: {
    quantity: {
      control: { type: 'number', min: 1, max: 100 },
      description: 'The current quantity value',
    },
    unit: {
      control: { type: 'select' },
      options: ['pcs', 'kgs', 'box', 'pack'],
      description: 'The current unit type',
    },
    size: {
      control: { type: 'select' },
      options: ['small', 'medium', 'large'],
      description: 'The size variant of the component',
    },
    onQuantityChange: {
      action: 'quantity-changed',
      description: 'Callback when quantity value changes',
    },
    onUnitChange: {
      action: 'unit-changed',
      description: 'Callback when unit type changes',
    },
  },
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof BuyingUnit>

export const Default: Story = {
  args: {
    quantity: 1,
    unit: 'pcs',
    onQuantityChange: action('quantity-changed'),
    onUnitChange: action('unit-changed'),
    size: 'medium',
  },
}

export const Small: Story = {
  args: {
    ...Default.args,
    size: 'small',
  },
}

export const Large: Story = {
  args: {
    ...Default.args,
    size: 'large',
  },
}

export const WithKilogramsUnit: Story = {
  args: {
    ...Default.args,
    quantity: 2,
    unit: 'kgs',
  },
}

export const WithBoxUnit: Story = {
  args: {
    ...Default.args,
    quantity: 5,
    unit: 'box',
  },
}

export const HighQuantity: Story = {
  args: {
    ...Default.args,
    quantity: 25,
    unit: 'pack',
  },
}

// Interactive story with user events
export const InteractiveTest: Story = {
  args: {
    ...Default.args,
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement)

    // Test quantity input change
    const quantityInput = canvas.getByLabelText('Quantity')
    await userEvent.clear(quantityInput)
    await userEvent.type(quantityInput, '5')

    // Test unit dropdown
    const unitButton = canvas.getByRole('button', { expanded: false })
    await userEvent.click(unitButton)

    // Wait for dropdown to open
    const dropdown = canvas.getByRole('listbox')
    expect(dropdown).toBeVisible()

    // Select different unit
    const kgsOption = canvas.getByText('kgs')
    await userEvent.click(kgsOption)

    // Verify dropdown closes
    expect(dropdown).not.toBeVisible()
  },
}

// Accessibility testing
export const AccessibilityTest: Story = {
  args: {
    ...Default.args,
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement)

    // Test keyboard navigation
    const quantityInput = canvas.getByLabelText('Quantity')
    const unitButton = canvas.getByRole('button')

    // Check ARIA attributes
    expect(quantityInput).toHaveAttribute('aria-label', 'Quantity')
    expect(unitButton).toHaveAttribute('aria-haspopup', 'listbox')
    expect(unitButton).toHaveAttribute('aria-expanded', 'false')

    // Test keyboard navigation
    await userEvent.tab()
    expect(quantityInput).toHaveFocus()

    await userEvent.tab()
    expect(unitButton).toHaveFocus()

    // Test Enter key on dropdown
    await userEvent.keyboard('{Enter}')
    expect(unitButton).toHaveAttribute('aria-expanded', 'true')
  },
}
