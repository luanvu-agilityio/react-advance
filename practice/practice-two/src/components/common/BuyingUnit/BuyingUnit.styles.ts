import styled from 'styled-components'

export const BuyingUnitContainer = styled.div<{ size?: string }>`
  display: inline-flex;
  align-items: center;
  background: var(--black-shade-6);
  border: 1px solid var(--black-shade-3);
  padding: ${(prop) => {
    switch (prop.size) {
      case 'small':
        return '6px'
      case 'large':
        return '18px 22px'

      default:
        return '14px 16px'
    }
  }};
  border-radius: 12px;
  height: ${(props) => {
    switch (props.size) {
      case 'small':
        return '32px'
      case 'large':
        return '56px'
      default:
        return '48px'
    }
  }};
  max-width: ${(props) => (props.size === 'small' ? '96px' : '111px')};
`

export const QuantityInput = styled.input<{ size?: string }>`
  width: ${(props) => (props.size === 'small' ? '34px' : '25px')};
  border: none;
  text-align: center;
  font-size: 15px;
  padding-right: ${(props) => (props.size === 'small' ? '6.5px' : '16px')};
  color: var(--black-color-default);
  background: var(--black-shade-6);

  &:focus {
    outline: none;
  }

  /* Hide number input arrows */
  &::-webkit-outer-spin-button,
  &::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }
  &[type='number'] {
    -moz-appearance: textfield;
  }
`

export const Separator = styled.div<{ size?: string }>`
  width: 1px;
  height: ${(props) => (props.size === 'small' ? '20px' : '24px')};
  background-color: var(--black-shade-3);
`

export const SelectContainer = styled.div`
  position: relative;
`

export const SelectButton = styled.button<{ size?: string }>`
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 4px;
  font-family: var(--font-family-primary);
  font-size: 12px;
  font-weight: var(--font-weight-semibold);
  padding-left: ${(props) => (props.size === 'small' ? '6.5px' : '15px')};
  color: var(--black-color-default);
  text-transform: capitalize;
  background: transparent;
  border: none;
  height: 100%;
  cursor: pointer;

  &:focus {
    outline: none;
  }
`

export const DropdownList = styled.ul<{ $isOpen: boolean }>`
  position: absolute;
  top: calc(100% + 8px);
  left: 0;
  right: 0;
  background: white;
  border: 1px solid var(--black-shade-4);
  border-radius: 8px;
  padding: 8px 0;
  display: ${(props) => (props.$isOpen ? 'block' : 'none')};
  box-shadow: 0 2px 10px rgba(0, 0, 0, 0.1);
  z-index: 10;
`

export const DropdownItem = styled.li`
  padding: 8px 16px;
  cursor: pointer;
  font-family: var(--font-family-primary);
  font-size: 12px;
  font-weight: var(--font-weight-semibold);
  color: var(--black-color-default);
  text-transform: capitalize;
  list-style: none;

  &:hover {
    background-color: var(--black-shade-5);
  }
`
