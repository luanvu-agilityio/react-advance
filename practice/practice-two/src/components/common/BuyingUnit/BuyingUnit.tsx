'use client'

import { useEffect, useRef, useState, type ChangeEvent } from 'react'
import { ChevronDown } from 'lucide-react'
import {
  BuyingUnitContainer,
  DropdownItem,
  DropdownList,
  QuantityInput,
  SelectButton,
  SelectContainer,
  Separator,
} from './BuyingUnit.styles'

interface BuyingUnitProps {
  quantity: number
  unit: string
  onQuantityChange: (value: number) => void
  onUnitChange: (value: string) => void
  size?: 'small' | 'medium' | 'large'
}

const units = ['pcs', 'kgs', 'box', 'pack']

const BuyingUnit = ({
  quantity,
  unit,
  onQuantityChange,
  onUnitChange,
  size = 'medium',
}: BuyingUnitProps) => {
  const [isOpen, setIsOpen] = useState(false)
  const [inputValue, setInputValue] = useState(quantity.toString())
  const dropdownRef = useRef<HTMLDivElement>(null)

  useEffect(() => {
    setInputValue(quantity.toString())
  }, [quantity])

  // Close dropdown when clicking outside
  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (
        dropdownRef.current &&
        !dropdownRef.current.contains(event.target as Node)
      ) {
        setIsOpen(false)
      }
    }

    document.addEventListener('mousedown', handleClickOutside)
    return () => {
      document.removeEventListener('mousedown', handleClickOutside)
    }
  }, [])

  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.target.value)

    const value = parseInt(e.target.value)
    if (!isNaN(value) && value > 0) {
      onQuantityChange(value)
    }
  }

  return (
    <BuyingUnitContainer size={size}>
      <QuantityInput
        type="number"
        min="1"
        value={inputValue}
        onChange={handleInputChange}
        aria-label="Quantity"
        size={size}
      />
      <Separator size={size} />
      <SelectContainer ref={dropdownRef}>
        <SelectButton
          onClick={() => setIsOpen(!isOpen)}
          aria-haspopup="listbox"
          aria-expanded={isOpen}
          size={size}
        >
          {unit}
          <ChevronDown size={12} />
        </SelectButton>
        <DropdownList $isOpen={isOpen}>
          {units.map((option) => (
            <DropdownItem
              key={option}
              aria-selected={option === unit}
              onClick={() => {
                onUnitChange(option)
                setIsOpen(false)
              }}
            >
              {option}
            </DropdownItem>
          ))}
        </DropdownList>
      </SelectContainer>
    </BuyingUnitContainer>
  )
}

export default BuyingUnit
