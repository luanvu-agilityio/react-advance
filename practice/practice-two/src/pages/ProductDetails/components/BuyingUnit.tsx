import { useEffect, useRef, useState, type ChangeEvent } from 'react'
import styled from 'styled-components'
import { ChevronDown } from 'lucide-react'

interface BuyingUnitProps {
  quantity: number
  unit: string
  onQuantityChange: (value: number) => void
  onUnitChange: (value: string) => void
  size?: 'small' | 'medium' | 'large'
}

const BuyingUnitContainer = styled.div<{ size?: string }>`
  display: inline-flex;
  align-items: center;
  background: var(--black-shade-6);
  border: 1px solid var(--black-shade-3);
  padding: ${(prop) => {
    switch (prop.size) {
      case 'small':
        return '6px'
      case 'large':
        return '18px 22px'

      default:
        return '14px 16px'
    }
  }};
  border-radius: 12px;
  height: ${(props) => {
    switch (props.size) {
      case 'small':
        return '32px'
      case 'large':
        return '56px'
      default:
        return '48px'
    }
  }};
  max-width: ${(props) => (props.size === 'small' ? '96px' : '111px')};
`

const QuantityInput = styled.input<{ size?: string }>`
  width: ${(props) => (props.size === 'small' ? '34px' : '25px')};
  border: none;
  text-align: center;
  font-size: 15px;
  padding-right: ${(props) => (props.size === 'small' ? '6.5px' : '16px')};
  color: var(--black-color-default);
  background: var(--black-shade-6);

  &:focus {
    outline: none;
  }

  /* Hide number input arrows */
  &::-webkit-outer-spin-button,
  &::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }
  &[type='number'] {
    -moz-appearance: textfield;
  }
`

const Separator = styled.div<{ size?: string }>`
  width: 1px;
  height: ${(props) => (props.size === 'small' ? '20px' : '24px')};
  background-color: var(--black-shade-3);
`

const SelectContainer = styled.div`
  position: relative;
`

const SelectButton = styled.button<{ size?: string }>`
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 4px;
  font-family: var(--font-family-primary);
  font-size: 12px;
  font-weight: var(--font-weight-semibold);
  padding-left: ${(props) => (props.size === 'small' ? '6.5px' : '15px')};
  color: var(--black-color-default);
  text-transform: capitalize;
  background: transparent;
  border: none;
  height: 100%;
  cursor: pointer;

  &:focus {
    outline: none;
  }
`

const DropdownList = styled.ul<{ $isOpen: boolean }>`
  position: absolute;
  top: calc(100% + 8px);
  left: 0;
  right: 0;
  background: white;
  border: 1px solid var(--black-shade-4);
  border-radius: 8px;
  padding: 8px 0;
  display: ${(props) => (props.$isOpen ? 'block' : 'none')};
  box-shadow: 0 2px 10px rgba(0, 0, 0, 0.1);
  z-index: 10;
`

const DropdownItem = styled.li`
  padding: 8px 16px;
  cursor: pointer;
  font-family: var(--font-family-primary);
  font-size: 12px;
  font-weight: var(--font-weight-semibold);
  color: var(--black-color-default);
  text-transform: capitalize;
  list-style: none;

  &:hover {
    background-color: var(--black-shade-5);
  }
`

const units = ['pcs', 'kgs', 'box', 'pack']

const BuyingUnit = ({
  quantity,
  unit,
  onQuantityChange,
  onUnitChange,
  size = 'medium',
}: BuyingUnitProps) => {
  const [isOpen, setIsOpen] = useState(false)
  const [inputValue, setInputValue] = useState(quantity.toString())
  const dropdownRef = useRef<HTMLDivElement>(null)

  useEffect(() => {
    setInputValue(quantity.toString())
  }, [quantity])

  // Close dropdown when clicking outside
  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (
        dropdownRef.current &&
        !dropdownRef.current.contains(event.target as Node)
      ) {
        setIsOpen(false)
      }
    }

    document.addEventListener('mousedown', handleClickOutside)
    return () => {
      document.removeEventListener('mousedown', handleClickOutside)
    }
  }, [])

  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.target.value)

    const value = parseInt(e.target.value)
    if (!isNaN(value) && value > 0) {
      onQuantityChange(value)
    }
  }

  const handleInputBlur = () => {
    const value = parseInt(inputValue)
    if (isNaN(value) ?? value < 1) {
      setInputValue('1')
      onQuantityChange(1)
    }
  }

  return (
    <BuyingUnitContainer size={size}>
      <QuantityInput
        type="number"
        min="1"
        value={inputValue}
        onChange={handleInputChange}
        onBlur={handleInputBlur}
        aria-label="Quantity"
        size={size}
      />
      <Separator size={size} />
      <SelectContainer ref={dropdownRef}>
        <SelectButton
          onClick={() => setIsOpen(!isOpen)}
          aria-haspopup="listbox"
          aria-expanded={isOpen}
          size={size}
        >
          {unit}
          <ChevronDown size={12} />
        </SelectButton>
        <DropdownList $isOpen={isOpen}>
          {units.map((option) => (
            <DropdownItem
              key={option}
              aria-selected={option === unit}
              onClick={() => {
                onUnitChange(option)
                setIsOpen(false)
              }}
            >
              {option}
            </DropdownItem>
          ))}
        </DropdownList>
      </SelectContainer>
    </BuyingUnitContainer>
  )
}

export default BuyingUnit
