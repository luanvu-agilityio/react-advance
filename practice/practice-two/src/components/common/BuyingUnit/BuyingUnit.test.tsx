<<<<<<< HEAD
import { render, screen, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
=======
import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { vi, describe, it, expect, beforeEach } from 'vitest'
>>>>>>> 7a35d9791a5da6fe80ff0a8541efaf78233de04d
import BuyingUnit from './BuyingUnit'

const defaultProps = {
  quantity: 1,
  unit: 'pcs',
<<<<<<< HEAD
  onQuantityChange: jest.fn(),
  onUnitChange: jest.fn(),
=======
  onQuantityChange: vi.fn(),
  onUnitChange: vi.fn(),
>>>>>>> 7a35d9791a5da6fe80ff0a8541efaf78233de04d
}

describe('BuyingUnit Component', () => {
  beforeEach(() => {
<<<<<<< HEAD
    jest.clearAllMocks()
  })

  // SNAPSHOT TESTS FOR RENDERING
  it('matches snapshots for different states', () => {
    // Default state
    const { container: defaultContainer } = render(
      <BuyingUnit {...defaultProps} />
    )
    expect(defaultContainer).toMatchSnapshot('default')

    // Different quantity and unit
    const { container: customContainer } = render(
      <BuyingUnit {...defaultProps} quantity={5} unit="kgs" />
    )
    expect(customContainer).toMatchSnapshot('custom-values')

    // Small size
    const { container: smallContainer } = render(
      <BuyingUnit {...defaultProps} size="small" />
    )
    expect(smallContainer).toMatchSnapshot('small-size')

    // Large size
    const { container: largeContainer } = render(
      <BuyingUnit {...defaultProps} size="large" />
    )
    expect(largeContainer).toMatchSnapshot('large-size')
  })

  // INTERACTIVE BEHAVIOR TESTS
  describe('Quantity Input Behavior', () => {
=======
    vi.clearAllMocks()
  })

  describe('Rendering', () => {
    it('renders with default props', () => {
      render(<BuyingUnit {...defaultProps} />)

      expect(screen.getByDisplayValue('1')).toBeInTheDocument()
      expect(screen.getByText('pcs')).toBeInTheDocument()
    })

    it('renders with custom quantity and unit', () => {
      render(<BuyingUnit {...defaultProps} quantity={5} unit="kgs" />)

      expect(screen.getByDisplayValue('5')).toBeInTheDocument()
      expect(screen.getByText('kgs')).toBeInTheDocument()
    })
  })

  describe('Quantity Input', () => {
>>>>>>> 7a35d9791a5da6fe80ff0a8541efaf78233de04d
    it('calls onQuantityChange when valid number is entered', async () => {
      const user = userEvent.setup()
      render(<BuyingUnit {...defaultProps} />)

      const input = screen.getByRole('spinbutton')
      await user.clear(input)
      await user.type(input, '5')

      expect(defaultProps.onQuantityChange).toHaveBeenCalledWith(5)
    })

    it('does not call onQuantityChange for invalid input', async () => {
      const user = userEvent.setup()
      render(<BuyingUnit {...defaultProps} />)

      const input = screen.getByRole('spinbutton')
      await user.clear(input)
      await user.type(input, 'abc')

      expect(defaultProps.onQuantityChange).not.toHaveBeenCalled()
    })
<<<<<<< HEAD
  })

  describe('Dropdown Behavior', () => {
    it('toggles dropdown visibility when clicking the select button', async () => {
      render(<BuyingUnit {...defaultProps} />)

      // Check dropdown is initially closed/hidden
      const dropdownList = screen.getByTestId('dropdown-list')
      expect(dropdownList).not.toHaveClass('$isOpen')

      // Open dropdown
      const selectButton = screen.getByRole('button', { name: /pcs/i })
      fireEvent.click(selectButton)

      // Check dropdown is open
      expect(dropdownList).toBeVisible()

      // Close dropdown by clicking button again
      fireEvent.click(selectButton)

      // Verify dropdown is closed
      expect(dropdownList).not.toHaveClass('$isOpen')
    })

    it('calls onUnitChange when selecting a unit option', () => {
      render(<BuyingUnit {...defaultProps} />)

      // Open dropdown
      const selectButton = screen.getByRole('button')
      fireEvent.click(selectButton)

      // Select a different option
      const kgsOption = screen.getByText('kgs')
      fireEvent.click(kgsOption)

      // Verify callback was called correctly
      expect(defaultProps.onUnitChange).toHaveBeenCalledWith('kgs')

      // Verify dropdown closed after selection
      const dropdownList = screen.getByTestId('dropdown-list')
      expect(dropdownList).not.toHaveClass('$isOpen')
    })

    it('closes dropdown when clicking outside', () => {
      render(
        <div>
          <BuyingUnit {...defaultProps} />
          <div data-testid="outside-area">Outside Area</div>
        </div>
      )

      // Open dropdown
      const selectButton = screen.getByRole('button')
      fireEvent.click(selectButton)

      // Verify dropdown is open
      const dropdownList = screen.getByTestId('dropdown-list')
      expect(dropdownList).toBeVisible()

      // Simulate click outside
      const outsideArea = screen.getByTestId('outside-area')
      fireEvent.mouseDown(outsideArea)

      // Verify dropdown closed
      expect(dropdownList).not.toHaveClass('$isOpen')
=======

    it('updates input value when quantity prop changes', () => {
      const { rerender } = render(<BuyingUnit {...defaultProps} quantity={1} />)

      rerender(<BuyingUnit {...defaultProps} quantity={10} />)

      expect(screen.getByDisplayValue('10')).toBeInTheDocument()
    })
  })

  describe('Unit Dropdown', () => {
    it('opens dropdown when button is clicked', async () => {
      const user = userEvent.setup()
      render(<BuyingUnit {...defaultProps} />)

      const button = screen.getByRole('button')
      await user.click(button)

      expect(screen.getByRole('listbox')).toBeVisible()
      expect(button).toHaveAttribute('aria-expanded', 'true')
    })

    it('closes dropdown when option is selected', async () => {
      const user = userEvent.setup()
      render(<BuyingUnit {...defaultProps} />)

      const button = screen.getByRole('button')
      await user.click(button)

      const kgsOption = screen.getByText('kgs')
      await user.click(kgsOption)

      expect(defaultProps.onUnitChange).toHaveBeenCalledWith('kgs')
      expect(screen.queryByRole('listbox')).not.toBeInTheDocument()
    })

    it('displays all available units', async () => {
      const user = userEvent.setup()
      render(<BuyingUnit {...defaultProps} />)

      const button = screen.getByRole('button')
      await user.click(button)

      expect(screen.getByText('pcs')).toBeInTheDocument()
      expect(screen.getByText('kgs')).toBeInTheDocument()
      expect(screen.getByText('box')).toBeInTheDocument()
      expect(screen.getByText('pack')).toBeInTheDocument()
    })

    it('closes dropdown when clicking outside', async () => {
      const user = userEvent.setup()
      render(
        <div>
          <BuyingUnit {...defaultProps} />
          <div data-testid="outside">Outside element</div>
        </div>
      )

      const button = screen.getByRole('button')
      await user.click(button)

      expect(screen.getByRole('listbox')).toBeVisible()

      const outsideElement = screen.getByTestId('outside')
      await user.click(outsideElement)

      await waitFor(() => {
        expect(screen.queryByRole('listbox')).not.toBeInTheDocument()
      })
    })
  })

  describe('Accessibility', () => {
    it('has proper ARIA attributes', () => {
      render(<BuyingUnit {...defaultProps} />)

      const input = screen.getByRole('spinbutton')
      const button = screen.getByRole('button')

      expect(input).toHaveAttribute('aria-label', 'Quantity')
      expect(button).toHaveAttribute('aria-haspopup', 'listbox')
      expect(button).toHaveAttribute('aria-expanded', 'false')
    })

    it('has proper keyboard navigation', async () => {
      const user = userEvent.setup()
      render(<BuyingUnit {...defaultProps} />)

      // Tab to input
      await user.tab()
      expect(screen.getByRole('spinbutton')).toHaveFocus()

      // Tab to button
      await user.tab()
      expect(screen.getByRole('button')).toHaveFocus()

      // Enter to open dropdown
      await user.keyboard('{Enter}')
      expect(screen.getByRole('listbox')).toBeVisible()
    })
  })

  describe('Edge Cases', () => {
    it('handles negative numbers correctly', async () => {
      const user = userEvent.setup()
      render(<BuyingUnit {...defaultProps} />)

      const input = screen.getByRole('spinbutton')
      await user.clear(input)
      await user.type(input, '-5')
      await user.tab()

      expect(input).toHaveValue(1)
      expect(defaultProps.onQuantityChange).toHaveBeenCalledWith(1)
    })

    it('handles decimal numbers correctly', async () => {
      const user = userEvent.setup()
      render(<BuyingUnit {...defaultProps} />)

      const input = screen.getByRole('spinbutton')
      await user.clear(input)
      await user.type(input, '2.5')

      expect(defaultProps.onQuantityChange).toHaveBeenCalledWith(2)
    })

    it('handles empty input correctly', async () => {
      const user = userEvent.setup()
      render(<BuyingUnit {...defaultProps} />)

      const input = screen.getByRole('spinbutton')
      await user.clear(input)
      await user.tab()

      expect(input).toHaveValue(1)
      expect(defaultProps.onQuantityChange).toHaveBeenCalledWith(1)
>>>>>>> 7a35d9791a5da6fe80ff0a8541efaf78233de04d
    })
  })
})
