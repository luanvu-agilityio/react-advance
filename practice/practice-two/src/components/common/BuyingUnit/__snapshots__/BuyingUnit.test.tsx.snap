// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuyingUnit Component matches snapshots for different states: custom-values 1`] = `
<div>
  <div
    class="sc-Qotzb kIlQEZ"
  >
    <input
      aria-label="Quantity"
      class="sc-fYsHOw eIEOpm"
      min="1"
      type="number"
      value="5"
    />
    <div
      class="sc-dKREkF llQWUV"
    />
    <div
      class="sc-fWnslK jEoszk"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        class="sc-dIMoHT ceyjfN"
      >
        kgs
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-down"
          fill="none"
          height="12"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m6 9 6 6 6-6"
          />
        </svg>
      </button>
      <ul
        class="sc-iQQCXo kMmmIP"
        data-testid="dropdown-list"
      >
        <li
          aria-selected="false"
          class="sc-gDpztx hmkFAg"
        >
          pcs
        </li>
        <li
          aria-selected="true"
          class="sc-gDpztx hmkFAg"
        >
          kgs
        </li>
        <li
          aria-selected="false"
          class="sc-gDpztx hmkFAg"
        >
          box
        </li>
        <li
          aria-selected="false"
          class="sc-gDpztx hmkFAg"
        >
          pack
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`BuyingUnit Component matches snapshots for different states: default 1`] = `
<div>
  <div
    class="sc-Qotzb kIlQEZ"
  >
    <input
      aria-label="Quantity"
      class="sc-fYsHOw eIEOpm"
      min="1"
      type="number"
      value="1"
    />
    <div
      class="sc-dKREkF llQWUV"
    />
    <div
      class="sc-fWnslK jEoszk"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        class="sc-dIMoHT ceyjfN"
      >
        pcs
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-down"
          fill="none"
          height="12"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m6 9 6 6 6-6"
          />
        </svg>
      </button>
      <ul
        class="sc-iQQCXo kMmmIP"
        data-testid="dropdown-list"
      >
        <li
          aria-selected="true"
          class="sc-gDpztx hmkFAg"
        >
          pcs
        </li>
        <li
          aria-selected="false"
          class="sc-gDpztx hmkFAg"
        >
          kgs
        </li>
        <li
          aria-selected="false"
          class="sc-gDpztx hmkFAg"
        >
          box
        </li>
        <li
          aria-selected="false"
          class="sc-gDpztx hmkFAg"
        >
          pack
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`BuyingUnit Component matches snapshots for different states: large-size 1`] = `
<div>
  <div
    class="sc-Qotzb kIlQEZ"
  >
    <input
      aria-label="Quantity"
      class="sc-fYsHOw eIEOpm"
      min="1"
      type="number"
      value="1"
    />
    <div
      class="sc-dKREkF llQWUV"
    />
    <div
      class="sc-fWnslK jEoszk"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        class="sc-dIMoHT ceyjfN"
      >
        pcs
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-down"
          fill="none"
          height="12"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m6 9 6 6 6-6"
          />
        </svg>
      </button>
      <ul
        class="sc-iQQCXo kMmmIP"
        data-testid="dropdown-list"
      >
        <li
          aria-selected="true"
          class="sc-gDpztx hmkFAg"
        >
          pcs
        </li>
        <li
          aria-selected="false"
          class="sc-gDpztx hmkFAg"
        >
          kgs
        </li>
        <li
          aria-selected="false"
          class="sc-gDpztx hmkFAg"
        >
          box
        </li>
        <li
          aria-selected="false"
          class="sc-gDpztx hmkFAg"
        >
          pack
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`BuyingUnit Component matches snapshots for different states: small-size 1`] = `
<div>
  <div
    class="sc-Qotzb ghwjbr"
  >
    <input
      aria-label="Quantity"
      class="sc-fYsHOw eQyTyc"
      min="1"
      type="number"
      value="1"
    />
    <div
      class="sc-dKREkF hrrFRx"
    />
    <div
      class="sc-fWnslK jEoszk"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        class="sc-dIMoHT iodpGc"
      >
        pcs
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-down"
          fill="none"
          height="12"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m6 9 6 6 6-6"
          />
        </svg>
      </button>
      <ul
        class="sc-iQQCXo kMmmIP"
        data-testid="dropdown-list"
      >
        <li
          aria-selected="true"
          class="sc-gDpztx hmkFAg"
        >
          pcs
        </li>
        <li
          aria-selected="false"
          class="sc-gDpztx hmkFAg"
        >
          kgs
        </li>
        <li
          aria-selected="false"
          class="sc-gDpztx hmkFAg"
        >
          box
        </li>
        <li
          aria-selected="false"
          class="sc-gDpztx hmkFAg"
        >
          pack
        </li>
      </ul>
    </div>
  </div>
</div>
`;
